pkg/operator/controller/dnsnameresolver/resolver.go

```diff
@@ -0,0 +1,340 @@
+				// needs to be updated.
+			case <-timer.C:
+				// After every tick, DNS lookup needs to be performed for nextDNSName, if it is not empty.
+				// The nextDNSName and timeTillNextLookup also needs to be updated.
+				if len(nextDNSName) > 0 {
+					if err := res.lookupDNSName(nextDNSName, numIPs); err != nil {
+						logrus.Warning(err)
+					}
+				}
+			case deletedDNSName = <-res.deleted:
+				// If the deleted DNS name is the DNS name that will be looked up next, then
+				// the nextDNSName and timeTillNextLookup needs to be updated. If another
+				// DNS name is deleted then nothing is needed to be done.
+				if nextDNSName != deletedDNSName {
+					continue
+				}
+			}
+			// Get the DNS name whose TTL will expire first. Based on the nextLookupTime for the nextDNSName, the timeTillNextLookup
+			// will be updated appropriately.
+			nextDNSName, nextLookupTime, numIPs, exists = res.getNextDNSNameDetails()
+			remainingDuration := time.Until(nextLookupTime)
+			if !exists || remainingDuration > defaultMaxTTL {
+				// If no DNS name is found OR If the remaining duration is greater than default maximum TTL, then perform DNS lookup
+				// after default maximum TTL.
+				timeTillNextLookup = defaultMaxTTL
+			} else if remainingDuration.Seconds() > 0 {
+				// If the remaining duration is positive and less than default maximum TTL, then perform DNS lookup
+				// after the remaining duration.
+				timeTillNextLookup = remainingDuration
+			} else {
+				// TTL of the DNS name has already expired, so send DNS lookup request as soon as possible.
+				timeTillNextLookup = 1 * time.Millisecond
+			}
+			timer.Reset(timeTillNextLookup)
+		}
+	}()
+}
+
+// Add is called whenever a DNSNameResolver object is added or updated.
+func (res *Resolver) Add(dnsName string, resolvedAddresses []networkv1alpha1.DNSNameResolverResolvedAddress, matchesRegular bool, objName string) {
+	res.dnsLock.Lock()
+	defer res.dnsLock.Unlock()
+
+	// Get the details of the resolved name corresponding to the DNS name, if it exists.
+	resolvedName, exists := res.dnsNames[dnsName]
+
+	if len(resolvedAddresses) == 0 {
+		// If no IP address is currently associated with the DNS name and the corresponding
+		// resolved name details also do not exist, then set the next lookup time for the
+		// DNS name as the default maximum TTL. Spawn a goroutine to send a DNS lookup
+		// request for the DNS name.
+		if !exists {
+			resolvedName.minNextLookupTime = time.Now().Add(defaultMaxTTL)
+			go res.lookupDNSName(dnsName, 0)
+		}
+	} else {
+		// If some IP addresses are associated with the DNS name, then determine the minimum
+		// next lookup time among the associated IP addresses.
+		first := false
+		for _, resolvedAddress := range resolvedAddresses {
+			if !first ||
+				resolvedAddress.LastLookupTime.Time.Add(time.Second*time.Duration(resolvedAddress.TTLSeconds)).Before(resolvedName.minNextLookupTime) {
```

Comment:
```suggestion
                       needToLookupNow := resolvedAddress.LastLookupTime.Time.Add(time.Second*time.Duration(resolvedAddress.TTLSeconds)).Before(resolvedName.minNextLookupTime) 
			if !first || needToLookupNow {
```

Reply:
Moved to openshift/coredns-ocp-dnsnameresolver#2